### PR TITLE
fixes #21374 - audit host interfaces

### PR DESCRIPTION
**Pull request by timogoebel**
Adds auditing to host interfaces. This is required to e.g. see when the IP address of a host changes.

image

### CODE REVIEW

**Comment by ares** [comment]
Looking forward to test this one. I wonder if STI works correctly here, recently we fixed templates where we had this case. From looking at screens it looks like a great addition!

**Comment by timogoebel** [comment]
> I wonder if STI works correctly here

@ares: Me too, after I read your comment. I think it should work fine but if you have any hints or I missed anything: Please do let me know :-)

File: app/helpers/audits_helper.rb
Comment (ares) [inline app/helpers/audits_helper.rb]: this changes the icon for all other audits, I don't disagree with the change but is it the best icon to represent other audits?

before
before

after
after

File: app/models/concerns/audit_extensions.rb
Comment (ares) [inline app/models/concerns/audit_extensions.rb]: could we also get a definition for mac please?

File: app/models/concerns/audit_extensions.rb
Comment (ares) [inline app/models/concerns/audit_extensions.rb]: `interface = host.example.com` is slightly confusing, maybe better name would be interface_hostname or interface_fqdn?

File: app/models/nic/base.rb
Comment (ares) [inline app/models/nic/base.rb:65]: we should ignore password (bmc interfaces), ideally it should use the same trick as e.g. compute_resource/image//user, see
https://github.com/theforeman/foreman/blob/develop/app/models/user.rb#L15
and https://github.com/theforeman/foreman/commit/518a2593744a7cdeaf3088f7fcf595b7cf6bd6f0 for an example

in this case we could just ignore it and hope that https://github.com/collectiveidea/audited/pull/339 will be finished soon

Comment (ares) [reply app/models/nic/base.rb:65]: pass_exposed

**Review by ares — CHANGES_REQUESTED** [review]
Overall it's a great addition! I found few nits which should be easy to address.

One issue that I see is that if I create a nic without DNS name and identifier, the audit does not have a clickable name so I can't navigate to audit details. We should probably define auditale_name in Nic::Base that would fall back to some 'unnamed' string.

no_link

One bigger question is, should we differentiate interfaces types? Right now we store everything as `Nic::Base` so we can't search based on interface type. E.g. we can't find changes on BMC interfaces easily.

File: app/helpers/audits_helper.rb
Comment (ares) [reply app/helpers/audits_helper.rb]: perhaps @Rohoover could suggest better generic icon or confirm this one is ok

Comment (timogoebel) [reply app/helpers/audits_helper.rb]: Basically, I changed the icons to patternfly recommendations. If there are other suggestions, I'm happy to change them.

Comment (Rohoover) [reply app/helpers/audits_helper.rb]: Would it be worth aligning these icons to the section of the application it affected/was initiated in... so anything host related would use the host icon found in the menu, Content would use the content icon. Just a thought.

Comment (ares) [reply app/helpers/audits_helper.rb]: That's what it does but we have specific incons only for ~3 objects so far. This one is default if none specific icon was set. Sometimes there's not a good icon, e.g. how would we represent things like domain, foreman proxy, operating system etc.

Comment (Rohoover) [reply app/helpers/audits_helper.rb]: Well for Operating System, it's found in the Host menu, so perhaps use the Host icon until/unless we are able to find something else.

File: app/models/nic/base.rb
Comment (timogoebel) [reply app/models/nic/base.rb:65]: @ares: This is how this looks for me:

image

**Comment by timogoebel** [comment]
@ares: Thanks for the review. I think I addressed all your comments. Mind to take another look?

File: app/models/nic/base.rb
Comment (dLobatog) [reply app/models/nic/base.rb:65]: @timogoebel @ares Have you tried the latest version? I also see the password encrypted (like @timogoebel screenshot, or this if going from "no pass" to "some pass").
screenshot from 2017-11-13 10-56-32

File: app/views/audits/show.html.erb
Comment (dLobatog) [inline app/views/audits/show.html.erb]: I think just `btn-default` (grey instead of dark blue) suffices, clicking on "Associated host" isn't really the main action a user should be doing here, rather just going back to check other audits.

screenshot from 2017-11-13 11-00-25

cc @Rohoover your thoughts?

**Review by dLobatog — CHANGES_REQUESTED** [review]
@timogoebel IMO this worked pretty well after a few tests, and the new icons / labels for interface audits look slick 😄 

This won't pass tests without a rebase because of the FactoryGirl -> FactoryBot rename. After rebasing, I think it should be merged for 1.17. Any qualms @ares ?

**Comment by timogoebel** [comment]
* rebased
* changed the "associated host" button from primary to default
* empty values are not rendered as `[empty]`
image

File: app/helpers/audits_helper.rb
Comment (tbrisker) [inline app/helpers/audits_helper.rb]: Could you please combine this condition into the same if? i.e. 
```ruby
if audit.associated_id && !audit.associated_name.blank? && type_name != 'Interface'
```

Comment (timogoebel) [reply app/helpers/audits_helper.rb]: Ouch, sure.

**Comment by ares** [comment]
It looks like test failures are related, @timogoebel could you please take a look?

**Comment by timogoebel** [comment]
Fixed the tests.

```diff
diff --git a/app/helpers/audits_helper.rb b/app/helpers/audits_helper.rb
index d7c155a2b..ff3a8d496 100644
--- a/app/helpers/audits_helper.rb
+++ b/app/helpers/audits_helper.rb
@@ -32,7 +32,12 @@ module AuditsHelper
       when 'Puppet Class'
         (id_to_label audit.audited_changes.keys[0], audit.audited_changes.values[0]).to_s
       else
-        name = audit.auditable_name.blank? ? audit.revision.to_label : audit.auditable_name
+        name = if audit.auditable_name.blank?
+                 revision = audit.revision
+                 revision.try(:to_audit_label) || revision.to_label
+               else
+                 audit.auditable_name
+               end
         name += " / #{audit.associated_name}" if audit.associated_id && !audit.associated_name.blank? && type_name != 'Interface'
         name
     end
diff --git a/app/models/nic/base.rb b/app/models/nic/base.rb
index de82eb28f..95accd371 100644
--- a/app/models/nic/base.rb
+++ b/app/models/nic/base.rb
@@ -199,7 +199,7 @@ module Nic
       ip_value.present? && public_send(subnet_field).contains?(ip_value)
     end

-    def to_label
+    def to_audit_label
       return "#{name} (#{identifier})" if name.present? && identifier.present?
       return "#{mac} (#{identifier})" if mac.present? && identifier.present?
       [mac, name, identifier, _('Unnamed')].detect(&:present?)
```

**Review by dLobatog — APPROVED** [review]

**Comment by dLobatog** [comment]
@ares OK to merge this now then? Now it shows [empty] so I think it solves the problem you mentioned in your review.

**Comment by ares** [comment]
> @ares OK to merge this now then? Now it shows [empty] so I think it solves the problem you mentioned in your review.

no it is not OK, empty means the password was empty, but try to set some new password for BMC nic, see my screen, tested with latest version, to reproduce
1. have a host with bmc device
2. edit it's nic, set new password
3. save the host
4. see audit list

password

File: app/models/nic/base.rb
Comment (ares) [reply app/models/nic/base.rb:65]: I wonder how our steps differ, but with I'm still able to reproduce with the latest version, I believe we should fix it the same way as I linked in previous comment at least to be consistent with other passwords in other objects

File: app/helpers/audits_helper.rb
Comment (ares) [inline app/helpers/audits_helper.rb]: I suppose you added this method to avoid problems in orchestration tests, but not every model has `to_audit_label` method defined, try would fail with `NoMethodError` here, it does not render 500 since there's the rescue below. We should not use exceptions for program flow, I think it's better to first verify the object `respond_to?(:to_audit_label)` or define it in application record (does not feel right).

**Comment by timogoebel** [comment]
@ares: Thanks for testing this again. Unfortunately, I am unable to reproduce this.
In the screenshot you provided, the password is not shown as encrypted. It is for me.

image

The encrypted value should be filtered in the audit record. This is happening here:
https://github.com/theforeman/foreman/blob/1f6a9b2c84ae629dda55330b7a4d8f9ed4e948dc/app/models/concerns/audit_extensions.rb#L88-L97

I guess, that this is not working correctly in your development environment. But I have a feeling, this is worth a separate Redmine Issue & PR.

**Comment by ares** [comment]
ok, that's fair, please allow me some time to dig into this a bit deeper, I hope to get to it in 1 or 2 days, if not, I won't block this PR

**Review by ares — APPROVED** [review]

**Comment by ares** [comment]
Sorry for false positive, my encryption key was not correctly configured, I messed up my ENV variable setup. After I fixed it, I see it works fine. As an apologize, I'm going to merge this :-)

Thanks @timogoebel, very nice addition!

**Pull request merged**
